flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/datasource/transactions/xa/domain/TransactionId.java

```diff
@@ -0,0 +1,273 @@
+            throws IOException {
+        if (FORMAT_ID != formatId) {
+            throw new IOException(String.format("Xid formatId (%s) is not valid", formatId));
+        }
+
+        final DataInputDeserializer gid = new DataInputDeserializer(globalTransactionId);
+        byte[] jobIdBytes = readJobId(gid);
+        int subtaskId = gid.readInt();
+
+        final DataInputDeserializer branch = new DataInputDeserializer(branchQualifier);
+        int numberOfSubtasks = branch.readInt();
+        long checkpoint = branch.readLong();
+
+        return new TransactionId(jobIdBytes, subtaskId, numberOfSubtasks, checkpoint, 0, restored);
+    }
+
+    public static TransactionId deserialize(byte[] bytes) {
+        try {
+            final DataInputDeserializer in = new DataInputDeserializer(bytes);
+            byte[] jobIdBytes = readJobId(in);
+            int subtaskId = in.readInt();
+            int numberOfSubtasks = in.readInt();
+            long checkpoint = in.readLong();
+            int attempts = in.readInt();
+            return restore(jobIdBytes, subtaskId, numberOfSubtasks, checkpoint, attempts);
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(e);
+        }
+    }
+
+    private static byte[] readJobId(DataInputDeserializer in) throws IOException {
+        byte[] jobIdBytes = new byte[JobID.SIZE];
+        in.read(jobIdBytes);
+        return jobIdBytes;
+    }
+
+    public TransactionId copy() {
+        return new TransactionId(
+                jobId, subtaskId, numberOfSubtasks, checkpointId, attempts, restored);
+    }
+
+    public TransactionId withBranch(long checkpointId) {
+        return new TransactionId(
+                jobId, subtaskId, numberOfSubtasks, checkpointId, attempts, restored);
+    }
+
+    public TransactionId withAttemptsIncremented() {
+        return new TransactionId(
+                jobId, subtaskId, numberOfSubtasks, checkpointId, attempts + 1, restored);
+    }
+
+    public long getCheckpointId() {
+        return checkpointId;
+    }
+
+    public boolean getRestored() {
+        return restored;
+    }
+
+    public int getAttempts() {
+        return attempts;
+    }
+
+    public String getXidValue() {
+        return String.format(
+                "%s:%s:%s",
+                getFormatId(),
+                byteToHexString(getGlobalTransactionId()),
+                byteToHexString(getBranchQualifier()));
+    }
+
+    @Override
+    public int getFormatId() {
+        return FORMAT_ID;
+    }
+
+    @Override
+    public byte[] getGlobalTransactionId() {
+        try {
+            // globalTransactionId = job id + task index
+            final DataOutputSerializer out = new DataOutputSerializer(1);
+            out.write(jobId, 0, JobID.SIZE);
+            out.writeInt(subtaskId);
+            return out.getSharedBuffer();
+        } catch (IOException e) {
+            throw new FlinkRuntimeException(e.getLocalizedMessage());
```

Comment:
just a little puzzle.
Why not use `throw new FlinkRuntimeException(e);` ?

Reply:
changed.